### ISSUE TITLE
parallel::makeCluster() freezes RStudio on macOS with R 4.x

### ISSUE THREAD

**Issue by RLumSK** [rstudio/rstudio#6692]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.3.944
    OS Version      : macOS Catalina 10.15.4 
    R Version       : R-devel (2020-04-18 r78249)

### Steps to reproduce the problem

```r
cl <- parallel::makeCluster(1)
```
Same if the number of nodes is > 1.

### Describe the problem in detail

I have got a (so far) unpublished R package using MC simulations that suddenly did not work anymore. Every time when I tried to start the simulation RStudio went to 100 % CPU load and did nothing anymore. Sometimes, in particular, during the first seconds, I could press the 'stop' button to shutdown the R process. If I waited too long, I had to kill the RStudio process. I narrowed down the problem to the single code line shown above. However, plain R (in the terminal or the R GUI) does work without any problem. I did run the following combinations: 

R version | R Studio (1.3.938) |  R Studio (1.3.944) | R Studio (1.2.5042)
----------|-----------------|--------------------| -------------------|
R-devel   |  FREEZE | FREEZE | FREEZE | FREEZE | 
R 4.0 RC | NA | FREEZE | FREEZE |
R 3.6.3 | NA | OK  | NA

NA: not tested.

My best guess is that this issue is related to [recent changes](https://developer.r-project.org/Blog/public/2020/03/17/socket-connections-update/) in 'parallel' 
Important (and again), it works in the R terminal and the normal R GUI, so, therefore, I believe that something needs to be modified in RStudio (I might be wrong though). Side note: If I used the RStudio check and build functionality where I also test examples, RStudio works fine.

#### Update 2020-04-24

I realised that `FREEZE` is probably not correct, the default timeout for the connection between master and worker in `parallel::makeCluster()` is two minutes. If you wait the full time, R will throw the following error message : 

```
> cl <- parallel::makeCluster(2)
Error in makePSOCKcluster(names = spec, ...) : 
  Cluster setup failed. 2 of 2 workers failed to connect.
```

### Describe the behaviour you expected

Calling 
```r
cl <- parallel::makeCluster(1)
```
does not freeze RStudio or fails with the mentioned error message. 

### Update 2020-04-26, potential intermediate work-around

Since I had to continue testing with R 4.X and I wanted to use *RStudio**, I played a little bit with the parameters, and (at least on my machine), I figured out the following: 

If I use the parameter `setup_timeout` and set it to < 1 s, it still works. Example:

```r
cl <- parallel::makeCluster(10, setup_timeout = 0.5)
```

I tested up to 100 nodes, without any problem. Obviously this **is not** a fix, and it may be different on other machines (not to mention that one may want to have these longer timeouts). However, it may provide an intermediate solution until a fix can be provided. Values > 1 s consistently did not work for me, according to the manual entry for `parallel::makeCluster` the default is 2 min. 

- [X] I have read the guide to submitting good bug reports at https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports . 
- [X] I have installed the latest version of RStudio and confirmed that the issue still persists.
- [ ] If I am reporting a RStudio crash, I have included a diagnostics report. https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report
*Not applicable, it does not really crash.*
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@RLumSK Thank you for raising the issue! I can reproduce this in RStudio Desktop on MacOS Catalina 10.15.4, but not in RStudio Server on Red Hat 7.8. Similarly, if I press on the stop sign quickly enough, it stops, but if I wait about 15 seconds, pressing on it aborts the R session. 1.2 on Mac behaves a little differently, in that the stop sign does nothing. We'll review this as we continue development of RStudio.

**Comment by kevinushey** [comment]
Looks like another version of this issue:

https://github.com/rstudio/rstudio/issues/1997

Looks like R now calls `socketSelect(..., timeout = N)` when creating a PSOCK cluster:

https://github.com/wch/r-source/blob/37d0a8a7b7f75c1ca15f532be38597d80966048a/src/library/parallel/R/snowSOCK.R#L201-L202

**Comment by RLumSK** [comment]
@kevinushey Thanks for the quick response and feedback. Actually, before I did post this issue, I had looked into the issue you mention. For me, a connecting between these two issues was not so obvious, but clearly, you have better insight. The main reason why I flagged it was that, whatever it is, it prevents parallel jobs using 'parallel' on macOS from functioning in *RStudio*, so basically the issue (perhaps) now became more 'general'.

**Comment by kevinushey** [comment]
I think I agree. Given that this issue basically means that the `parallel` package now doesn't work at all with RStudio + R 4.0.0, we need to try and fix this for v1.3. (Unfortunately too late for the imminent v1.3 release, but hopefully as a follow-up patch release)

**Comment by jmcphers** [comment]
@gtritchie will look at this.

**Comment by kevinushey** [comment]
@gtritchie: You might consider looking at a solution that makes use of `.rs.registerReplaceHook()`; I am thinking we could inject our own hook for `socketSelect()` that manages the value of `R_wait_usec` before and after `socketSelect()` is called.

The other option of course is re-exploring https://github.com/rstudio/rstudio-pro/pull/155 but seeing if we can figure out why this stalls RPCs and if there's an alternate approach we can take there.

**Comment by RLumSK** [comment]
I updated my initial error report above. I don't know whether it helps, but if the timeout is set manually to < 1 s than it still works (on my machine).

**Comment by HenrikBengtsson** [comment]
I don't have macOS, so can't verify, but try with:
```r
cl <- parallel::makeCluster(2, setup_strategy = "sequential")
```

If this works, you could add the following to your `~/.Rprofile` until this is fixed in R/RStudio:
```r
## WORKAROUND: https://github.com/rstudio/rstudio/issues/6692
## Revert to 'sequential' setup of PSOCK cluster in RStudio Console on macOS and R 4.0.0
if (Sys.getenv("RSTUDIO") == "1" && !nzchar(Sys.getenv("RSTUDIO_TERM")) && 
    Sys.info()["sysname"] == "Darwin" && getRversion() == "4.0.0") {
  parallel:::setDefaultClusterOptions(setup_strategy = "sequential")
}
```

**Comment by storopoli** [comment]
@HenrikBengtsson 
This works under Catalina 10.15.4 and R-4.0.0

**Comment by jeroen** [comment]
@HenrikBengtsson thanks that solves the problem for me as well.

**Comment by RLumSK** [comment]
@HenrikBengtsson Thank you, this is an excellent and efficient intermediate solution. It also works for me.

**Comment by storopoli** [comment]
@HenrikBengtsson what moving from "parallel" to "sequential" does in terms of performance?

**Comment by jeroen** [comment]
@storopoli see this post: https://developer.r-project.org/Blog/public/2020/03/17/socket-connections-update/index.html

**Comment by serbinsh** [comment]
Also failing for me:

```
> cl <- parallel::makeCluster(1)
Error in makePSOCKcluster(names = spec, ...) : 
  Cluster setup failed. 1 worker of 1 failed to connect.
```

10.14.16

```
> sessionInfo()
R version 4.0.0 (2020-04-24)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6
```

**Comment by serbinsh** [comment]
Any updates on this?

**Comment by gtritchie** [comment]
Is being worked on, intent is to get it into a 1.3 bugfix release.

**Comment by serbinsh** [comment]
Awesome, thanks!

**Comment by agilebean** [comment]
@gtritchie as I'm working on publishing a paper, can you give a very rough estimate when this 1.3 bug release will be out? 
I am just wondering if I should revert to R3.6 as alternative.

**Comment by HenrikBengtsson** [comment]
@agilebean, see my https://github.com/rstudio/rstudio/issues/6692#issuecomment-619645114 above - it provides a simple workaround.

**Comment by gtritchie** [comment]
@agilebean No I don't know when I'll get this fixed, nor when the corresponding release will happen. Assume several weeks at the very least.

**Comment by kevinushey** [comment]
This will be alleviated in the next preview release (+ next daily builds) via the same workaround detailed in @HenrikBengtsson's comment. However, we'll leave this issue open so we can have a more appropriate fix for v1.4 (or later) if possible.

**Comment by jdtuck** [comment]
I currently am having a similar issue and can file a new one. On R 4.0.2 and RStudio Server I can execute `paralell::makeCluster()`, but when the cluster actually runs using `foreach`, it takes forever and does not use all the cores. I do not have this problem however in the terminal starting R. It seems to be isolated to RStudio.

**Comment by kevinushey** [comment]
@jdtuck can you please provide a reproducible example + output of `sessionInfo()`?

**Comment by jdtuck** [comment]
Example

I gave the example of some computational complexity so it truly a computational intensive process

```
library(doParallel)
cores = detectCores()-1
cl = makeCluster(cores)
registerDoParallel(cl)

N = 120
A = matrix(rexp(160000, rate=.1), ncol=400)
system.time(gam<-foreach(k = 1:N, .combine=cbind) %dopar% {
  out = svd(A)
  out$d
})

stopCluster(cl)
```

Timing in terminal R: 
```
 user  system elapsed
  0.243   0.037  11.072
```

Timing in RStudio
```
user  system elapsed 
  0.474   0.798 229.638 
```

It should be noted when I run it in the terminal watch the processes via `top` all processes of R are running at 100%, when I run it in RStudio and watch `top` all the processes are <10%. 

Output of `sessionInfo()` from RStudio
```
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Workstation 7.8 (Maipo)

Matrix products: default
BLAS/LAPACK: /usr/local/lib64/libopenblas.so.0.2

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] doParallel_1.0.15 iterators_1.0.12  foreach_1.5.0    

loaded via a namespace (and not attached):
[1] compiler_4.0.2   tools_4.0.2      codetools_0.2-16
```

Output from `sessionInfo()` from terminal R.
```
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Workstation 7.8 (Maipo)

Matrix products: default
BLAS:   /usr/local/lib64/R/lib/libRblas.so
LAPACK: /usr/local/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
[1] doParallel_1.0.15 iterators_1.0.12  foreach_1.5.0

loaded via a namespace (and not attached):
[1] compiler_4.0.2   codetools_0.2-16
``

I have been scratching my head for the past couple of days on why this could be, given all the issues that are being reported are on mac. Though seeing both sessionInfo, there is a difference on the reported BLAS. Why would this be?

**Comment by kevinushey** [comment]
Thank you for taking the time to provide a reproducible example!

Unfortunately, I'm not able to reproduce -- this runs quickly as expected both on Ubuntu 20.04 and macOS. Does `parallel::detectCores()` return the expected number of cores on your machine?

The only other thing that stands out is that RStudio is evidently using a different BLAS than R itself is, but OpenBLAS should normally be faster than R's BLAS.

**Comment by jdtuck** [comment]
Thank you, I fixed the BLAS so it was matching, for some reason R had an old libR.so from 4.0.0 compile on the system. Headbanging now to ensue.

**Comment by kevinushey** [comment]
The "more appropriate fix" for this is IMHO out-of-scope for v1.4, since ultimately the only cost we're paying is a slower cluster startup time.

**Comment by Mthrun** [comment]
Would it be possible to reset at least the timeout parameter for macos users automatically to "setup_timeout = 0.5".
From my perspective, this is the most annoying bug in R 4.0.x in Mac Os Catalina

**Comment by kevinushey** [comment]
This issue should be resolved in the latest [preview release](https://www.rstudio.com/products/rstudio/download/preview/) of RStudio. If you have time, can you try and verify whether the issue does indeed appear to be fixed in your case?

**Comment by Mthrun** [comment]
Thanks for the hint. I can confirm that now the function call works.

**Comment by zhanglj37** [comment]
Yeah, this issue has been fixed in Rstudio v1.3.1056 released on July, 15, 2020. So the code of HenrikBengtsson can be changes into 
```
## WORKAROUND: https://github.com/rstudio/rstudio/issues/6692
## Revert to 'sequential' setup of PSOCK cluster in RStudio v1.3.959 or lower on macOS with R v4.0.0 or higher
library(rstudioapi)
if (Sys.getenv("RSTUDIO") == "1" && !nzchar(Sys.getenv("RSTUDIO_TERM")) && 
    Sys.info()["sysname"] == "Darwin" && getRversion() >= "4.0.0") {
    if(versionInfo()$version < "1.3.1056"){
	parallel:::setDefaultClusterOptions(setup_strategy = "sequential")
}  
}
```

**Comment by mmahmoudian** [comment]
I just stumbled upon this issue and found this thread from [a StackOverflow post](https://stackoverflow.com/a/64471474/1613005).

I confirm that adding the following line in my session worked:
```r
parallel:::setDefaultClusterOptions(setup_strategy = "sequential")
```

Here is my session info:
```
─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.0.3 (2020-10-10)
 os       Manjaro Linux               
 system   x86_64, linux-gnu           
 ui       RStudio                     
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Europe/Helsinki             
 date     2021-02-24                  

─ Packages ───────────────────────────────────────────────────────────────────
 package     * version date       lib source        
 assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.0.1)
 cli           2.3.0   2021-01-31 [1] CRAN (R 4.0.3)
 codetools     0.2-18  2020-11-04 [1] CRAN (R 4.0.3)
 foreach       1.5.1   2020-10-15 [1] CRAN (R 4.0.3)
 glue          1.4.2   2020-08-27 [1] CRAN (R 4.0.1)
 iterators     1.0.13  2020-10-15 [1] CRAN (R 4.0.1)
 rlang         0.4.10  2020-12-30 [1] CRAN (R 4.0.3)
 sessioninfo * 1.1.1   2018-11-05 [1] CRAN (R 4.0.1)
 withr         2.4.1   2021-01-26 [1] CRAN (R 4.0.3)
 yaml          2.2.1   2020-02-01 [1] CRAN (R 4.0.1)

[1] /home/mehrad/R/x86_64-pc-linux-gnu-library/4.0
```

Here is my system info:
```
Rstudio: Version 1.4.1103
Operating System: Manjaro Linux
KDE Plasma Version: 5.20.5
KDE Frameworks Version: 5.78.0
Qt Version: 5.15.2
Kernel Version: 5.10.15-1-MANJARO
OS Type: 64-bit
Processors: 8 × Intel® Core™ i7-4770K CPU @ 3.50GHz
Memory: 15.5 GiB of RAM
Graphics Processor: GeForce GTX 1060 6GB/PCIe/SSE2
```

This is interesting, because same R code was working for me about a month ago on the same R version and sample packages versions. What changed is what ever that has been updated through the package manager of my Linux distro (none of which are R or R packages as I have my own code to compile R on my system if needed)

**Comment by kevinushey** [comment]
Newer releases of RStudio enforce the usage of a sequential setup strategy:

https://github.com/rstudio/rstudio/blob/30872f8ae6086bd06f0c6ac2481bfe40b4ee0014/src/cpp/session/modules/SessionPatches.R#L35-L41

so I would expect that this no longer reproduces in the latest releases.

**Comment by mmahmoudian** [comment]
@kevinushey is this already in the 1.4.1106? Because for sure it is not in 1.4.1103

**Comment by kevinushey** [comment]
The issue had only ever been reported on macOS, so we only ever fixed the issue for macOS. If it's occurring on Linux as well, then we'll have to re-open and broaden the fix.

**Comment by kevinushey** [comment]
Note: scoping fix in 1.4-JR to only expand the fix in https://github.com/rstudio/rstudio/blob/30872f8ae6086bd06f0c6ac2481bfe40b4ee0014/src/cpp/session/modules/SessionPatches.R#L35-L41 to Linux as well. (And perhaps we ought to just set it on Windows as well.)

**Comment by martindoublem** [comment]
Can confirm, I just found this issue on ArchLinux today as well!

**Comment by kevinushey** [comment]
Thanks; will be fixed in the next daily builds + next patch release.

**Comment by astayleraz** [comment]
I tested this with no issues following the steps provide in the following setups: 
Win 10 version 1.4.1640 R 4.0.3
Mac 10.15.7 version 1.4.1640 R 4.0.3
Ubuntu 20 version 1.4.1640 R 3.6.3

Closing issue.

**Comment by tfjaeger** [comment]
I have a similar issue with the newest R (4.0.5; same for 4.0.2), newest RStudio (1.4.1106) and all libraries updated:

> library("future")> cl <- makeClusterPSOCK(workers = 1, revtunnel = TRUE, outfile = "", verbose = TRUE)

`[local output] Workers: [n = 1] ‘localhost’
[local output] Base port: 11230
[local output] Creating node 1 of 1 ...
[local output] - setting up node[local output] - attempt #1 of 3
[local output] Starting worker #1 on ‘localhost’: '/Library/Frameworks/R.framework/Resources/bin/Rscript' --default-packages=datasets,utils,grDevices,graphics,stats,methods -e 'try(suppressWarnings(cat(Sys.getpid(),file="/var/folders/6n/gc2x63fn3sl1bsp1wxxn__lc0000gn/T//Rtmpn5Z7dE/worker.rank=1.parallelly.parent=5033.13a94e7c8472.pid")), silent = TRUE)' -e 'workRSOCK <- tryCatch(parallel:::.slaveRSOCK, error=function(e) parallel:::.workRSOCK); workRSOCK()' MASTER=localhost PORT=11230 OUT= TIMEOUT=2592000 XDR=FALSE
[local output] - Exit code of system() call: 0
[local output] Waiting for worker #1 on ‘localhost’ to connect back 
starting worker pid=7363 on localhost:11230 at 15:38:39.619
Killing worker process (PID 7363) if still alive
Worker (PID 7363) was successfully killed: TRUE
Failed to launch and connect to R worker on local machine ‘localhost’ from local machine ‘Ts-Mac-Pro.local’.
 * The error produced by socketConnection() was: ‘reached elapsed time limit’ (which suggests that the connection timeout of 120 seconds (argument 'connectTimeout') kicked in)
 * The localhost socket connection that failed to connect to the R worker used port 11230 using a communication timeout of 2592000 seconds and a connection timeout of 120 seconds. 
 * Worker launch call: '/Library/Frameworks/R.framework/Resources/bin/Rscript' --default-packages=datasets,utils,grDevices,graphics,stats,methods -e 'try(suppressWarnings(cat(Sys.getpid(),file="/var/folders/6n/gc2x63fn3sl1bsp1wxxn__lc0000gn/T//Rtmpn5Z7dE/worker.rank=1.parallelly.parent=5033.13a94e7c8472.pid")), silent = TRUE)' -e 'workRSOCK <- tryCatch(parallel:::.slaveRSOCK, error=function(e) parallel:::.workRSOCK); workRSOCK()' MASTER=localhost PORT=11230 OUT= TIMEOUT=2592000 XDR=FALSE.
  * Worker (PID 7363) was successfully killed: TRUE 
  * Troubleshooting suggestions:
     - Suggestion #1: Set 'outfile=NULL' to see output from worker.  
     - Suggestion #2: Inspect the content of log file ‘/var/folders/6n/gc2x63fn3sl1bsp1wxxn__lc0000gn/T//Rtmpn5Z7dE/parallelly_makeClusterPSOCK_13a93250acbc.log’ for .`

**Comment by kevinushey** [comment]
Does it reproduce in the latest [daily builds](https://dailies.rstudio.com)?

**Comment by tfjaeger** [comment]
I'm getting another error. I've tried downloading the daily build from the link you provided but each time MacOS complains the the file is damaged.

**Comment by tfjaeger** [comment]
I was able to get rid of the CylanceProtect issue. But the issue with the cluster persists in the latest daily build (1.5.17):

> cl <- future::makeClusterPSOCK(workers = 1, outfile = "", verbose = TRUE)
[local output] Workers: [n = 1] ‘localhost’
[local output] Base port: 11429
[local output] Creating node 1 of 1 ...
[local output] - setting up node
[local output] - attempt #1 of 3
Testing if worker's PID can be inferred: ‘'/Library/Frameworks/R.framework/Resources/bin/Rscript' -e 'try(suppressWarnings(cat(Sys.getpid(),file="/var/folders/6n/gc2x63fn3sl1bsp1wxxn__lc0000gn/T//RtmpccKAAE/worker.rank=1.parallelly.parent=7968.1f20433af893.pid")), silent = TRUE)' -e "file.exists('/var/folders/6n/gc2x63fn3sl1bsp1wxxn__lc0000gn/T//RtmpccKAAE/worker.rank=1.parallelly.parent=7968.1f20433af893.pid')"’
- Possible to infer worker's PID: TRUE
[local output] Starting worker #1 on ‘localhost’: '/Library/Frameworks/R.framework/Resources/bin/Rscript' --default-packages=datasets,utils,grDevices,graphics,stats,methods -e 'try(suppressWarnings(cat(Sys.getpid(),file="/var/folders/6n/gc2x63fn3sl1bsp1wxxn__lc0000gn/T//RtmpccKAAE/worker.rank=1.parallelly.parent=7968.1f20433af893.pid")), silent = TRUE)' -e 'workRSOCK <- tryCatch(parallel:::.slaveRSOCK, error=function(e) parallel:::.workRSOCK); workRSOCK()' MASTER=localhost PORT=11429 OUT= TIMEOUT=2592000 XDR=FALSE
[local output] - Exit code of system() call: 0
[local output] Waiting for worker #1 on ‘localhost’ to connect back
starting worker pid=8035 on localhost:11429 at 16:16:43.514

**Comment by tfjaeger** [comment]
After 8 hours of trying various things, I'm pretty sure that this was not an RStudio issue (apologies for posting here). I don't know why it started working again, but a combination of starting in safe mode and then doing a hardware check (pressing D while restarting) seems to have repaired what multiple previous restarts didn't. Posting here in case other Mac users experience the same issue. I hope that's ok (if not, I can delete the posts).

**Comment by tfjaeger** [comment]
It turns out the issue I describe above re-occurs whenever I run brms with multiple cores (or perhaps other parallel processes) through RStudio (either the most recent stable release or the daily build 1.5.17) and my computer (MacPro 2013, MacOS 10.13) goes into sleep mode. The sampling process completes but the prompt is not being returned. Somehow brm() does not recognize that the process has completed. 

Following such an event, any parallelization call (through RStudio or R) hangs, including cl <- parallel::makeCluster(1, setup_strategy = "sequential"). It does not freeze RStudio (I can press stop to interrupt the process). If I run cl <- future::makeClusterPSOCK(workers = 1, outfile = "", verbose = TRUE) I get the output posted in previous posts. 

Unfortunately restarting the Mac in safemode does not seem to fix the issue. The upshot is that I have no idea what 'fixed' the issue last time, and that I now again can't use multiple cores. Any pointers---including where else to ask for help---would be much appreciated.

**Comment by tfjaeger** [comment]
Update: apparently, writing 

`127.0.0.1 localhost`

into /etc/hosts does the trick. (thanks to a little note at the very end of https://www.javaer101.com/es/article/36653434.html). Removing this line makes the issue re-appear, adding it make it disappear. Does this mean that when RStudio hangs after going to sleep on a multicore process it somehow modifies the hosts file?!

**Comment by HenrikBengtsson** [comment]
@tfjaeger, when doing (*):

```r
cl <- parallelly::makeClusterPSOCK(workers = nbr_of_workers)
```

internally you end up with the same as:

```r
cl <- parallelly::makeClusterPSOCK(workers = rep("localhost", times = nbr_of_workers))
```

From your most recent comment, it sounds like your local machine didn't know what `localhost` is, and you had to manually add a mapping to `127.0.0.1`.  That's unusual, but I guess it happens.

An alternative to editing `/etc/hosts`, which not everyone has the admin rights to do, you can also do:

```r
cl <- parallelly::makeClusterPSOCK(workers = rep("127.0.0.1", times = nbr_of_workers))
```

The same trick should also work when you use **parallel**, e.g.

```r
cl <- parallel::makeCluster(workers = rep("127.0.0.1", times = nbr_of_workers))
```

(*) Note that `future::makeClusterPSOCK()` is just an "alias" for `parallelly::makeClusterPSOCK()`, where the implementation lives these days. You can still use both, but one day we might ask everyone so specify `parallelly::...`.

FWIW, I've added it to **parallelly**'s todo list to be able to control the default hostname via an option/env var (https://github.com/HenrikBengtsson/parallelly/issues/51).

**Comment by tfjaeger** [comment]
Thank you @HenrikBengtsson. I agree that it's highly unusual that /etc/hosts is empty (it wasn't). Bizarrely, this problem (empty hosts file) seems to be itself a consequence of RStudio freezing at the end of parallelized sampling in brms::brm (as in a call to brm() with multiple cores that compiles, samples, but then hangs after all chains have completed sampling). When that freeze is resolved by force killing RStudio on my Mac, this seems to result in some system unstability. When I then restart my Mac (OS 10.13) that seems to cause the hosts file to be empty afterwards.

While I have been able to run parallel now (with the host file restored), I still run into the exact same brm issue (sampling completed and then it is stuck, as if the main rsession doesn't recognize that the separate r session started for the purpose of parallelization have completed). In search of a solution, I just installed the cmdstanr backend. With this alternative backened, brm does not seem to get stuck at the end of sampling (cmdstanr doesn't seem to create multiple rsessions, so I suspect that this is the reason).

**Comment by kevinushey** [comment]
> Does this mean that when RStudio hangs after going to sleep on a multicore process it somehow modifies the hosts file?!

No, it does not.

> Thank you @HenrikBengtsson. I agree that it's highly unusual that /etc/hosts is empty (it wasn't). Bizarrely, this problem (empty hosts file) seems to be itself a consequence of RStudio freezing at the end of parallelized sampling in brms::brm (as in a call to brm() with multiple cores that compiles, samples, but then hangs after all chains have completed sampling). When that freeze is resolved by force killing RStudio on my Mac, this seems to result in some system unstability. When I then restart my Mac (OS 10.13) that seems to cause the hosts file to be empty afterwards.

I have no idea why this would be the case, but (unless I'm missing something) RStudio does not touch the `/etc/hosts` file.

**Comment by tfjaeger** [comment]
Well, fwiw, I just had the same issue recur on my laptop (MacOS 10.13.6, Macbook Pro 2017; RStudio 1.2.1335; R 3.6.0), and again adding localhost to /etc/hosts fixed the issue. I note that---like for my other computer---this issue was absent until recently, and occurred without a software update.

**Comment by HenrikBengtsson** [comment]
Just want to inquire about the plan going forward with this `setup_strategy ="parallel"` issue:

In the most recent version, you're falling back to `setup_strategy ="sequential"` in RStudio Console across the board regardless of operating system;

https://github.com/rstudio/rstudio/blob/98979ded0817dd56248d1a16983a9565a206af0c/src/cpp/session/modules/SessionPatches.R#L35-L40

1. @kevinushey, in your comment from 2021-01-24, it was only done for macOS users.  Were the reports from other OSes that causes you to do this?  (FWIW, the "parallel" strategy seems to work for me on Ubuntu 18.04 Linux in the RStudio Console.)

2. Do you have a good understanding of the underlying issue, and if so, is it fixable?  Do you anticipate that the RStudio Console will eventually support `setup_strategy ="parallel"`?  Is there a timeline/ETA, or is this tracked under "someday"?

**Comment by kevinushey** [comment]
https://github.com/rstudio/rstudio/issues/6692#issuecomment-785207763 mentioned it occurred on Manjaro Linux as well.

The underlying issue was a bug in R (https://bugs.r-project.org/bugzilla/show_bug.cgi?id=18119) which should be fixed in R-devel now.

My guess is that perhaps the issue only effects certain versions of the C standard library; e.g. perhaps most glibc versions are okay, but musl or other C standard library implementations don't accept the timespec struct produced by R.

**Comment by HenrikBengtsson** [comment]
Thxs. Ah, I see. So, there's nothing to be fixed in RStudio, other than rolling back the workaround patch when fixed in R.  

I didn't make the connection to that [bug fix in R-devel](https://github.com/wch/r-source/commit/62f63f5d0aded0e988eeb4ead92347561af76057) on 2021-06-10. Hopefully, it'll also make its way into R 4.1.1.

**Comment by HenrikBengtsson** [comment]
One more thing, just so I get it correct: From you https://bugs.r-project.org/bugzilla/show_bug.cgi?id=18119#c0 report, it looks like this is not specific to RStudio Console.  Is that correct?  If so, I don't understand why some report that the problem goes away when they try the same thing running R in the terminal, e.g. https://github.com/rstudio/rstudio/issues/1997#issue-289327682 and https://github.com/HenrikBengtsson/future/issues/511#issuecomment-859699876. So, I'm still a bit confused what's going on here and whether this workaround is needed also outside the RStudio Console.

**Comment by kevinushey** [comment]
The bug only affects R front-ends that set R_wait_usec; AFAIK the built-in R front-ends don't do that.

You can reproduce in a plain R console if you explicitly set R_wait_usec (via C++ code or similar; see my example in the ticket)

**Comment by HenrikBengtsson** [comment]
Got it. Thanks for clarifying. Very helpful.

**Comment by HenrikBengtsson** [comment]
> The bug only affects R front-ends that set R_wait_usec; AFAIK the built-in R front-ends don't do that.

FWIW, the **tcltk** package [may set `R_wait_usec`](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/tcltk/src/tcltk_unix.c#L82).  I just tracked down a case where this causes this problem on CRAN's macOS servers, which can be reproduced using [`rhub::check(platform="macos-highsierra-release-cran")`](https://builder.r-hub.io/status/teeny_0.1.0.tar.gz-06e316805d65407fa8daddee08ee5162) and on [GitHub Actions w/ R 4.1.0 on macOS](https://github.com/HenrikBengtsson/teeny/runs/3076390940?check_suite_focus=true):

```r
R version 4.1.0 (2021-05-18) -- "Camp Pontanezen"
Copyright (C) 2021 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin17.0 (64-bit)

...

> loadNamespace("tcltk")
<environment: namespace:tcltk>
Warning message:
In fun(libname, pkgname) : couldn't connect to display ":50"
> cl <- parallel::makePSOCKcluster(1, setup_timeout = 15.0)
Error in parallel::makePSOCKcluster(1, setup_timeout = 15) : 
  Cluster setup failed. 1 worker of 1 failed to connect.
Execution halted
```

There's another package test _without_ `loadNamespace("tcltk")` and that works just fine.

This **tcltk** "interaction" seems to only affect macOS.

**Comment by kevinushey** [comment]
@HenrikBengtsson Thank you for the investigation!

**Comment by squirreltoken** [comment]
Have the same issue. As my simulation depends on it entirely and I will fail my thesis If I cannot fix this problem I feel very lost and frustrated:

The problem: 

`cl <- makecluster(1)` in a neverending computation that does not resolve.
- After tow minutes we get a long error message:

- `Failed to launch and connect to R worker on local machine ‘localhost’ from local machine ‘IRMW1234’.
 * The error produced by socketConnection() was: ‘reached elapsed time limit’ (which suggests that the connection timeout of 120 seconds (argument 'connectTimeout') kicked in)
 * The localhost socket connection that failed to connect to the R worker used port 11085 using a communication timeout of 2592000 seconds and a connection timeout of 120 seconds.
 * Worker launch call: "C:/PROGRA~1/R/R-41~1.0/bin/x64/Rscript" --default-packages=datasets,utils,grDevices,graphics,stats,methods -e "try(suppressWarnings(cat(Sys.getpid(),file=\"C:/Users/heja/AppData/Local/Temp/RtmpgvXRoz/worker.rank=1.parallelly.parent=2760.ac82e7022df.pid\")), silent = TRUE)" -e "workRSOCK <- tryCatch(parallel:::.workRSOCK, error=function(e) parallel:::.slaveRSOCK); workRSOCK()" MASTER=IRMW1234 PORT=11085 OUT=/dev/null TIMEOUT=2592000 XDR=FALSE SETUPTIMEOUT=120 SETUPSTRATEGY=sequential.
 * Failed to kill loc`

- Today is Monday. It worked perfectly fine on Friday. On Saturday, the problem started. Nothintg changed in between that I can recall. This is the worst. 
- I allowed R Studio as well as R to be granted permission by the Windows Firewall (was mentioned a couple of time online)
- I looked into ect/hosts . There we have a a server named localhost.  However when I ping localhost it mentions another server "::1". Might this be a problem? I dont have admin rights and tried this with the IT department. I use a computer which runs on the the institutional network. 

Any help is much apprfeciated.

**Comment by squirreltoken** [comment]
I accidentally fixed it: FYI; There were multiple different Versions and R libraries installed on my account, which I used on different computers. After deleting all kinds if R versions found on my account and installing R on the Computer, not on my account again, it all works perfectly fine. Dont know more details but hope it helps.

**Comment by leedrake5** [comment]
Just here to chime in - this problem has plagued me for years, and still continues with R 4.3.2.
